operator/buffer/memory.go

```diff
@@ -29,36 +31,135 @@ func NewMemoryBufferConfig() *MemoryBufferConfig {
 	}
 }
 
-// Build builds a MemoryBufferConfig into a Buffer, loading any entries that were previously unflushed
-// back into memory
-func (c MemoryBufferConfig) Build(context operator.BuildContext, pluginID string) (Buffer, error) {
-	return &MemoryBuffer{}, nil
+// Build builds a MemoryBufferConfig into a Buffer
+func (c MemoryBufferConfig) Build(_ operator.BuildContext, pluginID string) (Buffer, error) {
+	return &MemoryBuffer{
+		pluginID:      pluginID,
+		buf:           make(chan *entry.Entry, c.MaxEntries),
+		maxChunkDelay: c.MaxChunkDelay.Raw(),
+		maxChunkSize:  c.MaxChunkSize,
+		closed:        false,
+	}, nil
 }
 
-// MemoryBuffer is a buffer that holds all entries in memory until Close() is called,
-// at which point it saves the entries into a database. It provides no guarantees about
-// lost entries if shut down uncleanly.
+// MemoryBuffer is a buffer that holds all entries in memory until Close() is called.
+// Once close is called all entries will be lost it is reccommended to call Drain before Close.
 type MemoryBuffer struct {
+	pluginID      string
+	buf           chan *entry.Entry
+	maxChunkDelay time.Duration
+	maxChunkSize  uint
+	closed        bool
+
+	// readLock ensures only one concurrent read/drain can process at a time.
+	// If multiple reads or a drain happen while another is processing it may cause
+	// an indefinite block or longer blocks to maxChunkDelay.
+	readLock sync.Mutex
 }
 
 // Add adds an entry onto the buffer.
 // Is a blocking call if the buffer is full
 func (m *MemoryBuffer) Add(ctx context.Context, e *entry.Entry) error {
-	return nil
+	// If buffer is closed don't allow this operation
+	if m.closed {
+		return ErrBufferedClosed
+	}
+
+	// Insert into buffer or error if context finishes before we can
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("ctx error adding to buffer: %w", ctx.Err())
+	case m.buf <- e:
+		return nil
+	}
 }
 
 // Read reads from the buffer.
 // Read will block until the there are MaxChunkSize entries or we have block as long as MachChunkDelay.
 func (m *MemoryBuffer) Read(ctx context.Context) ([]*entry.Entry, error) {
-	return nil, nil
+	// If buffer is closed don't allow this operation
+	if m.closed {
+		return nil, ErrBufferedClosed
+	}
+
+	m.readLock.Lock()
+	defer m.readLock.Unlock()
```

Comment:
What happens here if close is called and locks the mutex here until it's finished? I'm guessing we'd end up just returning an empty slice right? Should we check for a closed buffer after this lock, instead of before?

Reply:
Yeah the closed check should really be after the mutex any place it's used. It's the cleanest way to prevent the race condition.